### PR TITLE
Removed cuda memory allocation which was causing CUDA OOM

### PR DESCRIPTION
**Pull request by karan6181**
## Description of changes:
- Removed the CUDA memory allocation which was used to synchronize the `shuffle_seed` and `prefix_int` across ranks.
- Based on how randomness works, the initial value `shuffle_seed` or `prefix_int` would be same across all the ranks, hence, there is no need for `dist.broadcast` operation.
- User needs to set the seed for getting the deterministic behavior on single node and multi-node training which is the same behavior as before.

## Issue #, if available:
CO-1591

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [x] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [x] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#running-tests))
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

**Comment by dakinggg** [comment]
QQ: Is it actually guaranteed that `prefix_int` and `shuffle_seed` are the same across all ranks (including multi node)? I think this may only be true if the main training script seeds randomness.

**Comment by karan6181** [comment]
> QQ: Is it actually guaranteed that `prefix_int` and `shuffle_seed` are the same across all ranks (including multi node)? I think this may only be true if the main training script seeds randomness.

For single node with 8 GPUs, I have verified that the `prefix_int` and `shuffle_seed`  is same across all ranks by printing those variables from all the ranks. Multi-node testing is in-progress.

File: streaming/base/dataset.py [hunk 1]
```diff
@@ -168,7 +167,12 @@ def __init__(self,
             filename = self._download_file(basename)
         else:
             filename = os.path.join(local, self.split, basename)  # pyright: ignore
-        dist.barrier()
+
+        # Everyone waits for the file to become populated.
+        is_file_exist(filename, TICK, self.download_timeout,
+                      f'{filename} file took too long to download')
+
+        # dist.barrier()
```
Comment (dakinggg) [inline streaming/base/dataset.py:175]: delete commented out code

File: streaming/base/dataset.py [hunk 2]
```diff
@@ -168,7 +167,12 @@ def __init__(self,
             filename = self._download_file(basename)
         else:
             filename = os.path.join(local, self.split, basename)  # pyright: ignore
-        dist.barrier()
+
+        # Everyone waits for the file to become populated.
+        is_file_exist(filename, TICK, self.download_timeout,
```
Comment (dakinggg) [inline streaming/base/dataset.py:172]: suggest renaming this function to `wait_for_file_to_exist`

**Review by dakinggg — COMMENTED** [review]
As discussed offline, suggest seeding the randomness explicitly or creating an rng instance so as not to rely on randomness being the same across nodes/ranks

File: streaming/base/dataset.py [hunk 3]
```diff
@@ -152,9 +150,10 @@ def __init__(self,
         self.download_timeout = download_timeout
         self.validate_hash = validate_hash or None
 
-        if tdist.is_available() and not tdist.is_initialized() and torch.cuda.is_available() and \
-                'RANK' in os.environ:
-            tdist.init_process_group('nccl')
+        if self.download_retry < 0:
+            raise ValueError('parameter `download_retry` can never be negative.')
```
Comment (knighton) [inline streaming/base/dataset.py:154]: Capitalize

Maybe "Number of download retries must be nonnegative"?

File: streaming/base/dataset.py [hunk 4]
```diff
@@ -152,9 +150,10 @@ def __init__(self,
...
+        if self.download_timeout < 0:
```
Comment (knighton) [inline streaming/base/dataset.py:155]: <= 0

Maybe "Download timeout must be greater than zero seconds"

File: streaming/base/util.py [hunk 5]
```diff
@@ -18,3 +20,26 @@ def get_list_arg(text: str) -> List[str]:
         List[str]: Splits, if any.
     """
     return text.split(',') if text else []
+
+
+def is_file_exist(filename: str, sleep_time: float, timeout: float, err_msg: str) -> None:
```
Comment (knighton) [inline streaming/base/util.py:25]: sleep_time -> poll_interval?

File: streaming/base/dataset.py (hunk 4 above)
Comment (knighton) [reply streaming/base/dataset.py:155]: Or "Download timeout (in seconds) must be greater than zero"

File: streaming/base/util.py (hunk 5 above)
Comment (karan6181) [reply streaming/base/util.py]: Done in https://github.com/mosaicml/streaming/pull/103/commits/bb6649673cbd20eeec770ef30d57668b3ed8afab.

File: streaming/base/dataset.py (hunk 4 above)
Comment (karan6181) [reply streaming/base/dataset.py:155]: Done in https://github.com/mosaicml/streaming/pull/103/commits/bb6649673cbd20eeec770ef30d57668b3ed8afab. I kept the parameter name for the user to know what to change.

File: streaming/base/dataset.py (hunk 3 above)
Comment (karan6181) [reply streaming/base/dataset.py]: Done in https://github.com/mosaicml/streaming/pull/103/commits/bb6649673cbd20eeec770ef30d57668b3ed8afab. I kept the parameter name for the user to know what to change.

File: streaming/base/dataset.py (hunk 2 above)
Comment (karan6181) [reply streaming/base/dataset.py]: Done in https://github.com/mosaicml/streaming/pull/103/commits/bb6649673cbd20eeec770ef30d57668b3ed8afab.

File: streaming/base/dataset.py (hunk 1 above)
Comment (karan6181) [reply streaming/base/dataset.py]: Done in https://github.com/mosaicml/streaming/pull/103/commits/bb6649673cbd20eeec770ef30d57668b3ed8afab.

**Comment by karan6181** [comment]
> QQ: Is it actually guaranteed that `prefix_int` and `shuffle_seed` are the same across all ranks (including multi node)? I think this may only be true if the main training script seeds randomness.

Yes, as of now, for deterministic behavior, user needs to set the seeds. If the user sets the seeds, then the behavior is deterministic on single and multi-node.

**Comment by dakinggg** [comment]
Should we set a default shuffle seed? Help the user avoid surprising outcomes, like having a different shuffle on each node. (Fwiw, I don't think our examples in the examples repo explicitly set the shuffle seed)

**Comment by knighton** [comment]
Can we just broadcast seed over gloo or something? Non-deterministic is a bug

**Comment by dakinggg** [comment]
Is there a problem with setting a default shuffle seed? To avoid adding unnecessary distributed stuff. Iirc having a separate gloo process group on mcloud is non trivial.

**Comment by knighton** [comment]
Good points... sorry am in Guanajuato with terrible reception... cannot really field code reviews atm... defer to what y'all think is best

**Comment by abhi-mosaic** [comment]
Would also prefer to avoid calling `torch.distributed` ! Greatly reduces the number of potential bugs. A default seed seems ok with me, unless someone can think of potential bugs?

**Comment by karan6181** [comment]
> Should we set a default shuffle seed? Help the user avoid surprising outcomes, like having a different shuffle on each node. (Fwiw, I don't think our examples in the examples repo explicitly set the shuffle seed)

Yes, we can. As of today, we kept the `shuffle_seed` to `None` as a default value to get the data randomness behavior and if the user needs a deterministic behavior, then they would have to set the `seed` value in their code, although doesn't necessarily have to set `shuffle_seed` explicitly. If user is using composer, then they can directly add `composer.utils.reproducibility.seed_all(seed)` in their script. 

Currently, I don't think so of any application where user needs a data randomness behavior in every run, and if they would, then they can set the `shuffle_seed` to different values in every run to get the data randomness behavior. 

The commit https://github.com/mosaicml/streaming/pull/103/commits/a8ef59cc5203b07f2011d96c4fe06e17cd6cbc9f adds the default `shuffle_seed` value.

**Comment by karan6181** [comment]
> Would also prefer to avoid calling `torch.distributed` ! Greatly reduces the number of potential bugs. A default seed seems ok with me, unless someone can think of potential bugs?

Agree. Adding a `torch.distributed` calls might surfaces some unknown bugs. With having a specific or a default `shuffle_seed`, user can get a deterministic behavior and non-deterministic behavior with different seed value in every run.

File: streaming/base/dataset.py [hunk 6]
```diff
@@ -139,7 +137,7 @@ def __init__(self,
                  download_retry: int = 2,
                  download_timeout: float = 60,
                  validate_hash: Optional[str] = None,
-                 shuffle_seed: Optional[int] = None,
+                 shuffle_seed: Optional[int] = 9176,
```
Comment (dakinggg) [inline streaming/base/dataset.py:140]: Type is no longer optional I think

**Review by dakinggg — APPROVED** [review]
LGTM, thanks!

File: streaming/base/dataset.py (hunk 6 above)
Comment (karan6181) [reply streaming/base/dataset.py]: Done in https://github.com/mosaicml/streaming/pull/103/commits/69425859864b6df1340636f4eb65245f7ac84538.
